src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateEntry.cs

```diff
@@ -1,12 +1,14 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System.Collections.Generic;
+
 namespace Microsoft.AspNetCore.Mvc.ModelBinding
 {
     /// <summary>
     /// An entry in a <see cref="ModelStateDictionary"/>.
     /// </summary>
-    public class ModelStateEntry
+    public abstract class ModelStateEntry
```

Comment:
@rynowak is running this by @Eilon. We could potentially remove some of this complexity if we targeted a `IReadOnlyDictionary` and the only way for users to add content to the dictionary is via the specific APIs we expose.